From 02b34eab7365799d20126791cc06c37b27b5bd65 Mon Sep 17 00:00:00 2001
From: Kristian Evensen <kristian.evensen@gmail.com>
Date: Thu, 14 Jul 2016 15:09:51 +0200
Subject: [PATCH] Random MAC fix cdc_ether

Need RX fixup due to crappy behavior

Error check

Style

Error

Typo

Blabla
---
 drivers/net/usb/cdc_ether.c | 24 ++++++++++++++++++++++++
 1 file changed, 24 insertions(+)

diff --git a/drivers/net/usb/cdc_ether.c b/drivers/net/usb/cdc_ether.c
index 7cba2c3..a8ae5b2 100644
--- a/drivers/net/usb/cdc_ether.c
+++ b/drivers/net/usb/cdc_ether.c
@@ -428,10 +428,33 @@ int usbnet_cdc_bind(struct usbnet *dev, struct usb_interface *intf)
 		return status;
 	}
 
+	if (dev->net->dev_addr[0] & 0x02)
+		eth_hw_addr_random(dev->net);
+
 	return 0;
 }
 EXPORT_SYMBOL_GPL(usbnet_cdc_bind);
 
+/* Make sure packet have correct destination MAC address
+ *
+ * A firmware bug observed on some devices (ZTE MF910, MF823) is that the device
+ * sends packets with a static, bogus, random MAC address (event if device
+ * address has been updated). Always set MAC address to that of the device.
+ */
+static int usbnet_cdc_rx_fixup(struct usbnet *dev, struct sk_buff *skb)
+{
+	if (skb->len < ETH_HLEN)
+		return 1;
+
+	skb_reset_mac_header(skb);
+
+	if ((eth_hdr(skb)->h_dest[0] & 0x02) &&
+	    !ether_addr_equal(eth_hdr(skb)->h_dest, dev->net->dev_addr))
+		ether_addr_copy(eth_hdr(skb)->h_dest, dev->net->dev_addr);
+
+	return 1;
+}
+
 static const struct driver_info	cdc_info = {
 	.description =	"CDC Ethernet Device",
 	.flags =	FLAG_ETHER | FLAG_POINTTOPOINT,
@@ -440,6 +463,7 @@ static const struct driver_info	cdc_info = {
 	.status =	usbnet_cdc_status,
 	.set_rx_mode =	usbnet_cdc_update_filter,
 	.manage_power =	usbnet_manage_power,
+	.rx_fixup = usbnet_cdc_rx_fixup,
 };
 
 static const struct driver_info wwan_info = {
-- 
2.5.0

